### PR TITLE
Fedpara updated

### PR DESCRIPTION
**Pull request by yehias21**
## Issue
#2031 
### Description
Updated draft
## Remaining tasks
[ ] Finalizing personalized FedPara with femnist

### CODE REVIEW

**Review by jafermarq — COMMENTED** [review]
Hey @yehias21,

This is just a preliminar review. I think before moving into adding `FEMNIST` experiments (which you can find how other baselines integrated that dataset here: https://github.com/adap/flower/tree/main/baselines/fedmeta -- i.e. cloning from LEAF directly), it's preferable to finalise the `CIFAR-10` and `CIFAR-100` experiments.

* Would it be possible to present the line plots in the same format as in Figure 3 in the paper. That means showing both `VGG16_ori` and `VGG16_fedpara` curves with the x-axis showing communication costs (instead of rounds). 
* The `environment setup` section needs to be completed with the instructions to follow in order to construct the python environment. (likely you can borrow from what your collegues did for `FedExp`: https://github.com/Roeia99/flower/tree/FedExP/baselines/fedexp#environment-setup)
* please include all necessary packages in `pyproject.toml` and detail how i can run the experiments (I see you provide a `run.sh` but you should mention this in the README.md in the `Running the experiments` section).

**Comment by yehias21** [comment]
Hi Javier, I've addressed all the comments related to the Cifar baselines as mentioned. Please inform me if there are any additional matters to attend to.

File: baselines/fedpara/fedpara/server.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import torch
+from flwr.common import NDArrays, Scalar
+from hydra.utils import instantiate
+from omegaconf import DictConfig
+from torch.utils.data import DataLoader
+
+from fedpara.models import test
+
+
+def get_on_fit_config(hypearparams: Dict):
```
Comment (jafermarq) [inline baselines/fedpara/fedpara/server.py:15]: note i moved this function out of `main.py`

File: baselines/fedpara/fedpara/models.py [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+    print(f"Learning rate: {lr}")
+    criterion = torch.nn.CrossEntropyLoss()
+    optimizer = torch.optim.SGD(
+        net.parameters(),
+        lr=lr,
+        momentum=hyperparams["momentum"],
+        weight_decay=hyperparams["weight_decay"],
+    )
+    net.train()
+    for _ in tqdm(range(epochs), desc="Local Training ..."):
```
Comment (jafermarq) [inline baselines/fedpara/fedpara/models.py:337]: Maybe it's worth removing `tqdm` for the clients? (It progress bars don't looks so nice when multiple processes update their own...). Another option is to have a flag in the configs to toggle it on/off (but better keep it off by default)

**Review by jafermarq — COMMENTED** [review]
Hi @yehias21,

See some comments below, mostly minor. You'll see I have pushed also some changes. Those primarily fix some formatting, mypy and pylint issues. A couple of issues remain. Could you look into those? You can run the tests on your local machine by: (1) activating your environment, (2) `cd ..`; then, (3) `./dev/test-baseline.sh fedpara`.

File: baselines/fedpara/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,141 @@
+[build-system]
+requires = ["poetry-core>=1.4.0"]
+build-backend = "poetry.masonry.api"
+
+[tool.poetry]
+name = "fedpara" # <----- Ensure it matches the name of your baseline directory containing all the source code
+version = "1.0.0"
+description = "Flower Baselines"
+license = "Apache-2.0"
+authors = ["The Flower Authors <hello@flower.dev>"]
```
[the remark this answers is not in the document]
Comment (yehias21) [reply baselines/fedpara/pyproject.toml:10]: Done!

File: baselines/fedpara/README.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+poetry run python -m fedpara.main
+# Important configs
+- model.conv_type --> choose parameterization scheme: lowrank or original(normal weights)
+- dataset_config.partition --> choosing between non IID and IID scheme
+- model.ratio --> choosing the ratio (lambda) of number of parameters
+
+**Multi-runs**
+
+# To run fedpara for non-iid cifar 10 on vgg16 for lowrank and original schemes
+poetry run python -m fedpara.main --multirun model.conv_type=standard,lowrank
```
[the remark this answers is not in the document]
Comment (yehias21) [reply baselines/fedpara/README.md:125]: Done!

File: baselines/fedpara/README.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
+---
+title: "FedPara: Low-rank Hadamard Product for Communication-Efficient Federated Learning"
+url:  https://openreview.net/forum?id=d71n4ftoCBy
+labels: [image classification, personalization, low-rank training, tensor decomposition]
+dataset: [CIFAR10, CIFAR100, FEMNIST]
```
[the remark this answers is not in the document]
Comment (yehias21) [reply baselines/fedpara/README.md:5]: Done!

File: baselines/fedpara/fedpara/models.py (hunk 2 above)
Comment (yehias21) [reply baselines/fedpara/fedpara/models.py:337]: done!

File: baselines/fedpara/_static/Cifar100_iid.jpeg
[the remark this answers is not in the document]
Comment (yehias21) [reply baselines/fedpara/_static/Cifar100_iid.jpeg:1]: image
Done!

[the remark this answers is not in the document]
Comment (jafermarq) [reply baselines/fedpara/_static/Cifar100_iid.jpeg:1]: looks great!

File: baselines/fedpara/fedpara/main.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            f"{cfg.clients_per_round}",
+        ]
+    )
+
+    utils.plot_metric_from_history(
+        hist=history,
+        save_plot_path=save_path,
+        suffix=file_suffix,
+        cfg=cfg,
+        model_size=net_glob.model_size()[1],
```
[the remark this answers is not in the document]
Comment (yehias21) [reply baselines/fedpara/fedpara/main.py:94]: Done!

**Comment by yehias21** [comment]
Done! all review comments checked and tests passed 10/10

File: baselines/fedpara/pyproject.toml [hunk 7]
```diff
@@ -0,0 +1,141 @@
+[build-system]
+requires = ["poetry-core>=1.4.0"]
+build-backend = "poetry.masonry.api"
+
+[tool.poetry]
+name = "fedpara" # <----- Ensure it matches the name of your baseline directory containing all the source code
+version = "1.0.0"
+description = "Flower Baselines"
+license = "Apache-2.0"
+authors = ["Yahia Salaheldin Shaaban <> ", "Omar Mokhtar <>", "Roeia Amr <>"]
```
Comment (jafermarq) [inline baselines/fedpara/pyproject.toml:10]: I formatted this slightly. If you wish to provide your contact email please do so, else remove the `<>` part.

**Review by jafermarq — COMMENTED** [review]
Hi @yehias21, 

Just a small comment below in `pyproject.toml`. I also made some edits, primarily to the `README.md`.

I ran your CIFAR-10/100 experiments and all curves resulted very close to the ones you include in the README with the exception of the `model.conv_type=standard` CIFAR-10 iid setting. The results i get are more in line with those in the paper than what your plots show. For example, for CIFAR-10 non-iid crosses the 60% threshold approx after 180GB of comms (see Fig3d in the paper), I get something similar, but in your plot it seems this happens after just 75GB of communication. For the IID CIFAR-10 setting I observe a similar situation: paper and results after running the code show croshing the 70% threshold after 100GB of communication but your plot shows that can be achieve with 50GB of communication.  
Screenshot 2024-01-03 at 22 32 18

I wonder the setting you run for those experiments you made the plots from was slightly different for CIFAR-10 with `model.conv_type=standard` and that's why the communication costs seem to be about half. Also note that max comm costs shown in the plots is 175GB while when running the code it would be a bit over 350GB (which matches the logic of (2 x 200rounds x 16clients x 15.3M params x 4 / 1024) -- I multiply by 4 to "convert" to MB. 

Finally do you have an update about FEMNIST?

File: baselines/fedpara/fedpara/main.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+from fedpara.dataset import load_datasets
+from fedpara.utils import (
+    get_parameters,
+    save_results_as_pickle,
+    seed_everything,
+    set_client_state_save_path,
+)
+
+
+@hydra.main(config_path="conf", config_name="mnist", version_base=None)
```
Comment (jafermarq) [inline baselines/fedpara/fedpara/main.py:20]:
```suggestion
@hydra.main(config_path="conf", config_name="cifar10", version_base=None)
```

Comment (jafermarq) [reply baselines/fedpara/fedpara/main.py:20]: Let's make `cifar10` the default config?

File: baselines/fedpara/fedpara/main.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+    # 3. Define clients
+    # In this scheme the responsability of choosing the client is on the client manager
+
+    client_fn = client.gen_client_fn(
+        train_loaders=train_loaders,
+        test_loader=test_loader,
+        model=cfg.model,
+        num_epochs=cfg.num_epochs,
+        args={"device": cfg.client_device, "algorithm": cfg.algorithm},
```
Comment (jafermarq) [inline baselines/fedpara/fedpara/main.py:53]:
```suggestion
        args={"algorithm": cfg.algorithm},
```

Comment (jafermarq) [reply baselines/fedpara/fedpara/main.py:53]: it's better to not set the device of clients directly in the config. Instead, based on `client_resources.num_gpus` it will be set to use CUDA (if a GPU is available). This is how other baselines do it (e.g. in [FedProx](https://github.com/adap/flower/blob/d7be8fb64507a83968479782976b9458435f909a/baselines/fedprox/fedprox/client.py#L155). See my comment later in `get_client_fn`

File: baselines/fedpara/fedpara/dataset.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+                )
+                pickle.dump(train_datasets, open(ds_path, "wb"))
+                dataset_test = Dataset(
+                    data_directory, train=False, download=True, transform=transform_test
+                )
+
+    test_loader = DataLoader(dataset_test, batch_size=batch_size, num_workers=2)
+    train_loaders = [
+        DataLoader(
+            DatasetSplit(dataset_train, ids),
```
Comment (jafermarq) [inline baselines/fedpara/fedpara/dataset.py:117]: There seems too be a problem with `ids`. It is an integer but inside `DatasetSplit` you do `list()`. That raises an error....

File: baselines/fedpara/fedpara/main.py (hunk 9 above)
Comment (jafermarq) [reply baselines/fedpara/fedpara/main.py:53]: Also, `algorithm` is not found in all configs (e.g. in cifar10)... maybe this is just some on-going work for MNIST? I think the easy fix will be to do in `cifar10.yaml` and `cifar100.yaml` the same as you do in `mnist.yaml` regarding the algorithm

File: baselines/fedpara/fedpara/main.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+            ),
+        )
+
+    else:
+        evaluate_fn = server.gen_evaluate_fn(
+            num_clients=cfg.num_clients,
+            test_loader=test_loader,
+            model=cfg.model,
+            device=cfg.server_device,
+            state_path=cfg.state_path,
```
Comment (jafermarq) [inline baselines/fedpara/fedpara/main.py:85]: `state_path` is not present for CIFAR configs so this raises an error. I think the best would be to define `state_path` as:
```
state_path: null
```
for both `cifar10.yaml` and `cifar100.yaml` and then in `get_evaluate_fn()` (and other places if uses) ignore this. But inspecting `get_evaluate_fn` i see it doesn't accept a `state_path` argument.

Comment (jafermarq) [reply baselines/fedpara/fedpara/main.py:85]: so maybe we should remove that last argument ?

File: baselines/fedpara/README.md [hunk 12]
```diff
@@ -0,0 +1,165 @@
+---
+title: "FedPara: Low-rank Hadamard Product for Communication-Efficient Federated Learning"
+url:  https://openreview.net/forum?id=d71n4ftoCBy
+labels: [image classification, personalization, low-rank training, tensor decomposition]
+dataset: [CIFAR-10, CIFAR-100, FEMNIST]
```
Comment (jafermarq) [inline baselines/fedpara/README.md:5]:
```suggestion
dataset: [CIFAR-10, CIFAR-100, MNIST]
```

Comment (jafermarq) [reply baselines/fedpara/README.md:5]: please update the readme indicating that the personalisation dataset used is MNIST.

File: baselines/fedpara/README.md [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+| IID | Non-IID |
+|:----:|:----:|
+|![Cifar100 iid](_static/Cifar100_iid.jpeg) | ![Cifar100 non-iid](_static/Cifar100_noniid.jpeg) |
+
+
+### CIFAR-10 (Accuracy vs Communication Cost)
+
+| IID | Non-IID |
+|:----:|:----:|
+|![CIFAR10 iid](_static/Cifar10_iid.jpeg) | ![CIFAR10 non-iid](_static/Cifar10_noniid.jpeg) |
```
Comment (jafermarq) [inline baselines/fedpara/README.md:165]: At the end of the readme you can add a new section showing how the personalisation experiments with MNIST can be executed. Even if they are not fully working, this would give others a good starting point. But please add a comment right under the title of the section stating that it has been quite hard reproducing the results (maybe give some details of what you tried, what would be the next steps to investigate etc)

File: baselines/fedpara/fedpara/dataset.py (hunk 10 above)
Comment (jafermarq) [reply baselines/fedpara/fedpara/dataset.py:117]: Should it be replaced instead with `[ ]`?

File: baselines/fedpara/fedpara/models.py [hunk 14]
```diff
@@ -0,0 +1,528 @@
...
+import torch
+import torch.nn.functional as F
+from flwr.common import Scalar
+from torch import nn
+from torch.nn import init
+from torch.utils.data import DataLoader
+
+
+class LowRankNN(nn.Module):
+    """."""
```
Comment (jafermarq) [inline baselines/fedpara/fedpara/models.py:16]: Please update with some brief description.

File: baselines/fedpara/fedpara/models.py [hunk 15]
```diff
@@ -0,0 +1,528 @@
...
+        init.kaiming_normal_(self.Y, mode="fan_out", nonlinearity=activation)
+
+    def forward(self):
+        """Do forward."""
+        out = torch.einsum("yr,xr->yx", self.Y, self.X)
+        return out
+
+
+class Linear(nn.Module):
+    """."""
```
Comment (jafermarq) [inline baselines/fedpara/fedpara/models.py:45]: Please update with some brief description.

File: baselines/fedpara/fedpara/models.py [hunk 16]
```diff
@@ -0,0 +1,528 @@
...
+        if self.pfedpara:
+            w = self.w1() * self.w2() + self.w1()
+        else:
+            w = self.w1() * self.w2()
+        out = F.linear(x, w, self.bias)
+        return out
+
+
+class FC(nn.Module):
+    """."""
```
Comment (jafermarq) [inline baselines/fedpara/fedpara/models.py:93]: Please update with some brief description.

**Review by jafermarq — COMMENTED** [review]
Hi @yehias21 and team,

Following our last chat, i think we can start wrapping up this baseline and focus on CIFAR-10/100 results. It's fine leaving the personalization with MNIST ones as an extension (can be added as a new section at the end of the readme -- see my comments about that below).

I've pushed a commit that fixing formatting and some typing issues. Most issues are resolved except for a couple (one being passing an argument that's not expected to `gen_evaluate_fn()` and another in client's `_get_keys_state_dict()`)

Please find my comments below. I applied them all locally, the code runs but I'm not seeing the global model improving. I ran this:
```bash
python -m fedpara.main --multirun model.conv_type=standard,lowrank 
```
It could be that I made a mistake.

Let me know when you have gone through my comments.

**Review by jafermarq — APPROVED** [review]
Hi @yehias21 and collaborators! Your implementation of `FedPara` is ready to be merged! Amazing work!!
